packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -0,0 +1,570 @@
+  previous,
+  /// Selection ends in this [Selectable].
+  ///
+  /// Part of the [Selectable] may or may not be selected, but there are still
+  /// content to select forward or backward.
+  end,
+  /// The result can't be determined in this frame.
+  ///
+  /// This is typically used when the subtree is scrolling to reveal more content.
+  pending,
+  /// There is no result for the selection event.
+  ///
+  /// This is used when the a selection result is not applicable, e.g.
+  /// [SelectAllSelectionEvent], [ClearSelectionEvent], and
+  /// [SelectWordSelectionEvent].
+  none,
+}
+
+/// The abstract interface to handle selection events.
+///
+/// {@template flutter.rendering.SelectionHandler}
+/// This class returns a [SelectionGeometry] as its [value], and is responsible
+/// to notify its listener when its selection geometry has changed as the result
+/// of receiving selection events.
+/// {@endtemplate}
+abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
+  /// Marks this handler to be responsible for painting leader layers for the
+  /// selection handles.
+  ///
+  /// This handler is responsible for painting the leader layers with the
+  /// given layer links if they are not null. It is possible that only one layer
+  /// is non-null if this handler is only responsible for painting one layer
+  /// link.
+  ///
+  /// The `startHandle` needs to be placed at the visual location of selection
+  /// start, the `endHandle` needs to be placed at the visual location of selection
+  /// end
+  void pushHandleLayers(LayerLink? startHandle, LayerLink? endHandle);
+
+  /// Gets the selected content in this object.
+  ///
+  /// Return `null` if nothing is selected.
+  SelectedContent? getSelectedContent();
+
+  /// Handles the [SelectionEvent] sent to this object.
+  ///
+  /// The subclasses need to update their selections or delegate the
+  /// [SelectionEvent]s to its subtree.
+  ///
+  /// The `event`s are subclasses of [SelectionEvent]. Use runtime type check to
+  /// determine what kinds of event are dispatched to this handler and handle
+  /// them accordingly.
+  SelectionResult dispatchSelectionEvent(SelectionEvent event);
+}
+
+/// The selected content in a [Selectable].
+class SelectedContent {
+  /// Creates a selected content object.
+  ///
+  /// Only supports plain text.
+  const SelectedContent({required this.plainText});
+
+  /// The selected content in plain text format.
+  final String plainText;
+}
+
+/// A mixin that can be selected by users when under a [SelectionArea] widget.
+///
+/// This object receives selection events and the [value] must reflect the
+/// current selection in this [Selectable]. The object must also notify its
+/// listener if the [value] ever changes.
+///
+/// This object is responsible for drawing selection highlight.
+///
+/// In order to receive the selection event, the mixer need to register
+/// themselves to [SelectionRegistrar]s. Use the
```

Comment:
Nit: I think it's more clear to say "itself" instead of "themselves" to be clear that we're referring to a single mixer.